python/dgl/nn/pytorch/explain/pgexplainer.py

```diff
@@ -0,0 +1,356 @@
+            number of nodes, and :math:`D` is the feature size.
+        edge_mask : Tensor, optional
+            Learned importance mask of the edges in the graph, which is a tensor
+            of shape :math:`(E)`, where :math:`E` is the number of edges in the
+            graph. The values are within range :math:`(0, 1)`. The higher,
+            the more important. Default: None.
+        """
+        num_nodes, _ = feat.shape
+        num_edges = graph.num_edges()
+
+        init_bias = self.init_bias
+        std = nn.init.calculate_gain("relu") * math.sqrt(2.0 / (2 * num_nodes))
+
+        if edge_mask is None:
+            self.edge_mask = torch.randn(num_edges) * std + init_bias
+        else:
+            self.edge_mask = edge_mask
+
+        self.edge_mask = self.edge_mask.to(graph.device)
+
+    def clear_masks(self):
+        r"""Clear the edge mask that play a crucial role to explain the
+        prediction made by the GNN for a graph.
+        """
+        self.edge_mask = None
+
+    def loss(self, prob, ori_pred):
+        r"""The loss function that is used to learn the edge
+        distribution.
+
+        Parameters
+        ----------
+        prob: Tensor
+            Tensor contains a set of probabilities for each possible
+            class label of some model for all the batched graphs,
+            which is of shape :math:`(1, L)`, where :math:`L` is the
+            different types of label in the dataset.
+        ori_pred: Tensor
+            Tensor of shape ::math:`(1, 1)`, representing the original prediction
+            for the graph.
+
+        Returns
+        -------
+        float
+            The function that returns the sum of the three loss components,
+            which is a scalar tensor representing the total loss.
+        """
+        target_prob = prob.gather(-1, ori_pred.unsqueeze(-1))
+        # 1e-6 added to prob to avoid taking the logarithm of zero
+        target_prob += 1e-6
+        # computing the log likelihood for a single prediction
+        pred_loss = torch.mean(-torch.log(target_prob))
+
+        # size
+        edge_mask = self.sparse_mask_values
+        if self.coff_budget <= 0:
+            size_loss = self.coff_budget * torch.sum(edge_mask)
+        else:
+            size_loss = self.coff_budget * F.relu(
+                torch.sum(edge_mask) - self.coff_budget
+            )
+
+        # entropy
+        scale = 0.99
+        edge_mask = self.edge_mask * (2 * scale - 1.0) + (1.0 - scale)
+        mask_ent = -edge_mask * torch.log(edge_mask) - (
+            1 - edge_mask
+        ) * torch.log(1 - edge_mask)
+        mask_ent_loss = self.coff_connect * torch.mean(mask_ent)
+
+        loss = pred_loss + size_loss + mask_ent_loss
+        return loss
+
+    def concrete_sample(self, log_alpha, beta=1.0, training=True):
+        r"""Sample from the instantiation of concrete distribution when training.
+
+        Parameters
+        ----------
+        log_alpha : Tensor
```

Comment:
This name is a bit confusing. Call it logits or `w` as in the paper?

Reply:
Done.